src/pip/_internal/resolution/resolvelib/requirements.py

```diff
@@ -37,11 +48,14 @@ def is_satisfied_by(self, candidate: Candidate) -> bool:
         return candidate == self.candidate
 
 
-class SpecifierRequirement(Requirement):
+class SpecifierRequirement(Requirement, KeyBasedCompareMixin):
     def __init__(self, ireq: InstallRequirement) -> None:
         assert ireq.link is None, "This is a link, not a specifier"
         self._ireq = ireq
         self._extras = frozenset(canonicalize_name(e) for e in self._ireq.extras)
+        KeyBasedCompareMixin.__init__(
+            self, key=str(ireq), defining_class=SpecifierRequirement
+        )
```

Comment:
Instead of using `str` here we may want to make `InstallRequirement` hashable and comparable.

Reply:
Although this may change the logic.